### PR TITLE
Towards Roman Level 2 ASDF data product support

### LINKED ISSUE #1355: [FEAT] Imviz: Add Roman ASDF Support
### Description

Need to add support for Roman ASDF files. Currently imviz does not understand what an ASDF file is (it complains that it's not a FITS file).

### Additional context

[🐱](https://jira.stsci.edu/browse/JDAT-2478)

### PR DESCRIPTION
**Pull request by bmorris3**
#### Background

This PR implements basic support for Roman Level 2 data products in Imviz. This effort was begun by @tddesjardins in his prototyping PR #1361. I am further revising and testing a version of that prototype in this branch/PR. Many of the revisions I've made to the implementation by @tddesjardins came from comments in [a helpful review](https://github.com/spacetelescope/jdaviz/pull/1361#pullrequestreview-992461713) by @pllim in June.

#### Test data

The Roman team is producing an ever-updating set of synthetic data products. I wrote this PR around a specific set of synthetic data, which you can read more about in the collapsed section below:

<details>
  <summary>Old background on test data</summary>

#### Outdated: ~Test data~

I am working with simulated exposures in the Roman ASDF format, which are used by the concept notebook. Some notes on these simulated exposures:
* The uncalibrated, synthetic, WFI imaging-mode exposures were produced in the [`SOC Build 22Q4_B7` Test Suite](https://innerspace.stsci.edu/pages/viewpage.action?spaceKey=RI&title=SOC+Build+22Q4_B7+Test+Suite) (this link is only for STScI internal folks, sorry others!)
* The simulated exposures have stars, instrument artifacts, noise, realistic WCS, and more. None of the data is real, even the stars are randomly generated!
* @tddesjardins then kindly ran [romancal](https://github.com/spacetelescope/romancal) on the uncalibrated (Level 1) exposures to make the calibrated versions (Level 2)
* @tddesjardins confirmed that these files can be shared publicly, so I've put them on [Box](https://stsci.box.com/s/7lhsv0vltsgjxdy1opmhib7ci2431cac) for ease of experimenting. STScI folks: DM me if you want the shared paths for the files.
* Each WFI exposure produces 18 ASDF files, one per detector. Each file is 390.3 MB. The Box directory linked above contains "one exposure" across 18 detectors, and weighs 7 GB.~
</details>

The PR will now run tests on tiny synthetic `roman_datamodels.datamodels.ImageModel` objects (~20x20 pixels, plus metadata), instead of the gigantic, synthetic data products. 

#### Dependencies 

This PR adds [`roman_datamodels`](https://github.com/spacetelescope/roman_datamodels) as an optional dependency, and falls back on the `asdf` package if `roman_datamodels` is not available.

#### Demo

I added a concept notebook in `notebooks/concepts/ImvizExample-asdf.ipynb`, which is a variation on `ImvizExample.ipynb` adapted for the Roman ASDF L2 data files. Of course, I will plan to add more comprehensive documentation as POs and/or other devs request it. Though I'm not sure how urgent it is to add docs when no one outside of the Roman team knows how to procure the files that are supported by this PR 😅 .

#### Testing

I have not added any tests yet. I could use some ideas from other devs on how to approach testing. Perhaps it would it be wise to make a tiny version of a Roman ASDF file, perhaps one that just has a small subset of the `data` array, but all of the usual metadata. That file could be small enough to put on Box and add to the remote-data tests. Thoughts?

#### Speed

Testing with glue-core v1.6, I find the following load times for one 4k image (from a single WFI detector):
```
# load a single WFI detector image from ASDF
imviz.load_data:
    1.02 s ± 19.7 ms per loop

# alternate between removing and adding data
imviz.remove_data_from_viewer && imviz.add_data_to_viewer:
    123 ms ± 3 ms per loop
```

Fixes #1355
Supersedes / closes #1361
[🐱](https://jira.stsci.edu/browse/JDAT-3150)

### Change log entry

- [x] Is a change log needed? If yes, is it added to `CHANGES.rst`? If you want to avoid merge conflicts,
  list the proposed change log here for review and add to `CHANGES.rst` before merge. If no, maintainer
  should add a `no-changelog-entry-needed` label.

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [x] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [x] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Did the CI pass? If not, are the failures related?
- [x] Is a milestone set? Set this to bugfix milestone if this is a bug fix and needs to be released ASAP; otherwise, set this to the next major release milestone.
- [x] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

### CODE REVIEW

File: jdaviz/configs/imviz/plugins/parsers.py
Comment (pllim) [inline jdaviz/configs/imviz/plugins/parsers.py:60]: This broke the test using `download_file` because the file comes back as `contents` without any extension. I think we should keep the `else` fallback and assume FITS if nothing else matches.

Comment (pllim) [reply jdaviz/configs/imviz/plugins/parsers.py:60]: Unless you have a better way to figure out the filetype if the filename has no extension.

Comment (bmorris3) [reply jdaviz/configs/imviz/plugins/parsers.py:60]: Good point, @pllim. Do you think it would be appropriate to make a tiny PR to `astropy.utils.data` that allows a user to discover file extensions of local caches of remote data? I think I have a working example, so I may open an astropy PR today.

Comment (bmorris3) [reply jdaviz/configs/imviz/plugins/parsers.py:74]: I've implemented a possible workaround in b60872b. 🤞🏻

Comment (bmorris3) [reply jdaviz/configs/imviz/plugins/parsers.py:74]: True, all parsers are affected by this problem. If that's something we want to fix generally, should that effort get its own ticket/issue so it can get prioritized/pointed? I'm not sure it is a pressing problem yet.

Comment (pllim) [reply jdaviz/configs/imviz/plugins/parsers.py:74]: Yes, we can defer, but you need to put "assume FITS" back in `else` for this PR.

Comment (bmorris3) [reply jdaviz/configs/imviz/plugins/parsers.py:74]: Agreed! 👍🏻

File: jdaviz/configs/imviz/plugins/parsers.py
Comment (pllim) [inline jdaviz/configs/imviz/plugins/parsers.py:75]: What if someone passes in ASDF file that isn't Roman? Is there a way to identify that it is actually Roman data before attempting to read with Roman package?

File: jdaviz/configs/imviz/plugins/parsers.py
Comment (pllim) [inline jdaviz/configs/imviz/plugins/parsers.py:81]:
```suggestion
        else:  # Assume FITS
```

File: jdaviz/configs/imviz/plugins/parsers.py
Comment (pllim) [inline jdaviz/configs/imviz/plugins/parsers.py:296]: Does this account for all the observing modes or whatever that Roman will churn out?

File: jdaviz/configs/imviz/plugins/parsers.py
Comment (pllim) [inline jdaviz/configs/imviz/plugins/parsers.py:304]: Should this use Jesse's label generator?

File: jdaviz/configs/imviz/plugins/parsers.py
Comment (pllim) [inline jdaviz/configs/imviz/plugins/parsers.py:317]: Do you mean to add this new empty "section" or did you mean to replace the existing section below?

**Review by pllim — COMMENTED** [review]
When this is ready to go forward, will need at least one test (possibly multiple to cover all the logic paths) and user facing documentation and declaring new dep in `setup.cfg`.

Change log can wait for now.

Thanks!

File: jdaviz/configs/imviz/plugins/parsers.py
Comment (bmorris3) [reply jdaviz/configs/imviz/plugins/parsers.py:75]: The [`roman_datamodels` docs page on the structure of Roman data products](https://roman-datamodels.readthedocs.io/en/latest/roman_datamodels/datamodels/general_structure.html#overview) suggests that it is a "convention" that data are stored under a `roman` attribute. I've confirmed this in the L2 products. If you run:
```python
import asdf

with asdf.open(path) as f:
    f.info()
```
you'll see
```
root (AsdfObject)
├─asdf_library (Software)
│ ├─author (str): The ASDF Developers
│ ├─homepage (str): http://github.com/asdf-format/asdf
│ ├─name (str): asdf
│ └─version (str): 2.13.0
├─history (dict)
│ └─extensions (list) ...
└─roman (WfiImage) # The schema for WFI Level 2 images.
...
```
Maybe the right thing to do is (1) open the file with the `asdf` package, (2) check for a `roman` attribute and load with `roman_datamodels` if so, (3) otherwise try passing the ASDF file without `roman_datamodels`.

File: jdaviz/configs/imviz/plugins/parsers.py
Comment (bmorris3) [reply jdaviz/configs/imviz/plugins/parsers.py:296]: This line was contributed by @tddesjardins, and the background on the available extensions in the `roman_datamodels` docs is rather sparse ([e.g.: this page doesn't seem finished](https://roman-datamodels.readthedocs.io/en/latest/roman_datamodels/datamodels/models.html)). Maybe @tddesjardins can comment?

**Comment by bmorris3** [comment]
> declaring new dep in `setup.cfg`.

The `roman_datamodels` package is not a strict installation requirement, since we fall back on `asdf` if `rdd` is not available (see [`try: import` block here](https://github.com/spacetelescope/jdaviz/pull/1822/files#diff-6395a570d2b676bda465e4269d61f3742cba505bccddebc5af310747245aff5fR18-R22)). Should we require it in `setup.cfg` anyway?

**Comment by pllim** [comment]
Since it is in such early development, what about a new "roman" section here?

https://github.com/spacetelescope/jdaviz/blob/cbd66959b79175caa68d350695a323afeb09bd4a/setup.cfg#L53

File: docs/imviz/import_data.rst
Comment (pllim) [inline docs/imviz/import_data.rst:106]:
```suggestion
You can also load Nancy Grace Roman Space Telescope (hereafter, Roman) data products, which are
```

File: docs/imviz/import_data.rst
Comment (pllim) [inline docs/imviz/import_data.rst:107]:
```suggestion
provided as ASDF files. If an ASDF file has a ``roman`` attribute, Jdaviz will
```

File: setup.cfg
Comment (pllim) [inline setup.cfg:62]: asdf is already required.

```suggestion
```

File: setup.cfg
Comment (pllim) [inline setup.cfg:64]:
```suggestion
```

File: jdaviz/configs/imviz/plugins/parsers.py
Comment (bmorris3) [reply jdaviz/configs/imviz/plugins/parsers.py]: Fixed 👍🏻

File: jdaviz/configs/imviz/plugins/parsers.py
Comment (pllim) [reply jdaviz/configs/imviz/plugins/parsers.py:309]: Imviz has its own rules since it can load wildcard extension and have both FITS and ASDF support, and so on. If the function is hidden, doesn't hurt if you also want to pass in `app` to them, but becareful to not break current Imviz labeling behavior. 🤪

File: jdaviz/configs/imviz/tests/test_parser.py
Comment (bmorris3) [inline jdaviz/configs/imviz/tests/test_parser.py:27]: `roman_datamodels > 0.14.1` will have tools for generating synthetic WFI-like data products on the fly. I've added a switch to allow tests to run when a sufficient version of roman_datamodels is installed.

File: jdaviz/configs/imviz/plugins/parsers.py
Comment (bmorris3) [inline jdaviz/configs/imviz/plugins/parsers.py:330]: Currently the dev version of `roman_datamodels` has a method for generating WFI-like data products, but it doesn't generate artificial WCS yet. I've created an issue over there for requesting this: https://github.com/spacetelescope/roman_datamodels/issues/134.

**Comment by bmorris3** [comment]
Progress update: this PR was created when `roman_datamodels` was on v0.14.0, and was designed to successfully load a particular set of synthetic Roman data products – see the "Test data" heading in the PR description.

Since then, `roman_datamodels` has advanced, and as mentioned in https://github.com/spacetelescope/jdaviz/pull/1822#discussion_r1140622125, there are now some tools for generating synthetic WFI image data models that would be very useful for testing in jdaviz. But since the PR was designed around the old synthetic data products, and they are no longer compatible with `roman_datamodels >= v0.14.2`, we need to either pin `roman_datamodels` or drop support for the old synthetic data. Since approximately no one cares at this stage, I'll do the latter. 

As of [f92fcb2](https://github.com/spacetelescope/jdaviz/pull/1822/commits/f92fcb2d7bad74ad6265321b8e7ea2e00dc977a5), I'm beginning to develop tests based on the dev version of `roman_datamodels`, and I will drop support for the outdated test data in the next days.

File: jdaviz/conftest.py
Comment (bmorris3) [inline jdaviz/conftest.py:325]: This can be removed when https://github.com/spacetelescope/roman_datamodels/pull/138 is addressed.

File: setup.cfg
Comment (bmorris3) [inline setup.cfg:66]: For now, I'm requiring dev versions of `roman_datamodels` and `rad`. We can minpin to the next release whenever it's available.

File: jdaviz/conftest.py
Comment (bmorris3) [inline jdaviz/conftest.py:395]: This line is the main difference from the analogous method in `roman_datamodels.testing.factories.create_wfi_image`. We need a synthetic GWCS object to use in our example image file, so I adapted a photutils method for creating one, and attach it to the image model.

File: tox.ini
Comment (bmorris3) [inline tox.ini:3]: I've added a `romandeps` testing option, and a CI test that uses it (allowed_fail=True), but I'm not 100% sure I've done this sensibly. Tips welcome!

File: setup.cfg
Comment (pllim) [inline setup.cfg:65]: This is temporary, right? I don't think we can release to PyPI with a pin like this.

File: tox.ini
Comment (pllim) [inline tox.ini:60]: Redundant
```suggestion
```

File: .github/workflows/ci_workflows.yml
Comment (pllim) [inline .github/workflows/ci_workflows.yml:78]: I don't like adding roman tests in the CI like this. It is probably going to fail a lot at this stage, and that will make reviewing unrelated PRs a pain.

Comment (bmorris3) [reply .github/workflows/ci_workflows.yml:78]: I set `allow_failure: true` with this in mind. Should we leave the `romandeps` option in there, but not add it to the CI workflow?

File: setup.cfg
Comment (bmorris3) [reply setup.cfg:65]: Correct.

File: .github/workflows/ci_workflows.yml
Comment (pllim) [reply .github/workflows/ci_workflows.yml:78]: I think I need to understand the plan of this PR and timeline of merge first before I can answer.

File: .github/workflows/ci_cron_weekly.yml
Comment (pllim) [inline .github/workflows/ci_cron_weekly.yml:52]: I wonder if we should just merge this into the job above. What is the point of having two variants of dev test for weekly cron when we already run a bulk of dev test in PRs?

**Comment by pllim** [comment]
Please see https://github.com/bmorris3/jdaviz/pull/3 for my proposed changes. Thanks!

File: docs/conf.py
Comment (kecnry) [inline docs/conf.py:271]: oh fun, the repo and python package name have underscore, but on pip/RTD its a hyphen?  Is there anywhere else that might also need updates (`PYTEST_HEADER_MODULES` maybe)?

```suggestion
intersphinx_mapping['roman_datamodels'] = ('https://roman-datamodels.readthedocs.io/en/stable/', None)
```

File: jdaviz/configs/imviz/plugins/parsers.py
Comment (kecnry) [inline jdaviz/configs/imviz/plugins/parsers.py:4]: does it make sense to move `asdf` to the roman-optional dependencies as well?

File: jdaviz/configs/imviz/plugins/parsers.py
Comment (kecnry) [inline jdaviz/configs/imviz/plugins/parsers.py:83]: any chance this logic could break (roman appearing in `asdf_file` for another reason like observer's last name)?

File: docs/imviz/import_data.rst
Comment (kecnry) [inline docs/imviz/import_data.rst:111]: Should we suggest the `[roman]` option when installing jdaviz anywhere in the docs (or defer until more support is added)?

File: jdaviz/configs/imviz/tests/utils.py
Comment (braingram) [inline jdaviz/configs/imviz/tests/utils.py:241]: Would the `mk_level2_image` maker utility in roman datamodels be useful here?
https://github.com/spacetelescope/roman_datamodels/blob/main/src/roman_datamodels/maker_utils.py#L485

File: jdaviz/configs/imviz/tests/utils.py
Comment (cshanahan1) [inline jdaviz/configs/imviz/tests/utils.py:248]: why cant the factory function just be imported from roman_datamodels? i worry this would be hard to keep up to date as the files evolve

File: setup.cfg
Comment (cshanahan1) [inline setup.cfg:64]: if you haven't already, add a few lines to the readme and install instructions showing how to install with the optional roman stuff

Comment (cshanahan1) [reply setup.cfg:64]: also, im not sure how this works but if you define this additional section here, will something like pip install .['roman'] work automatically? i see that you've defined the 'romandeps' option, but is that redundant?

File: tox.ini
Comment (cshanahan1) [inline tox.ini:39]: should the dev version of roman datamodels be added to devdeps as well?

File: docs/conf.py
Comment (pllim) [reply docs/conf.py:271]: Huh, good catch! I think I updated it everywhere else but missed this one. Why is RTD not complaining? We have nitpicky

https://github.com/spacetelescope/jdaviz/blob/6a2892c1c321547be7ea11756fadb086387a49b3/docs/conf.py#L236

**Review by cshanahan1 — COMMENTED** [review]
did an initial pass an left some comments, im going to test it out mode in depth before i leave more

File: jdaviz/configs/imviz/plugins/parsers.py
Comment (pllim) [reply jdaviz/configs/imviz/plugins/parsers.py:4]: No, asdf is already a required dependency from a long time ago for JWST support.

https://github.com/spacetelescope/jdaviz/blob/6a2892c1c321547be7ea11756fadb086387a49b3/setup.cfg#L46

File: docs/imviz/import_data.rst
Comment (pllim) [reply docs/imviz/import_data.rst:111]: I'd say defer for now.

File: jdaviz/configs/imviz/tests/utils.py
Comment (pllim) [reply jdaviz/configs/imviz/tests/utils.py:241]: Maybe, I didn't know it existed. But I'll defer to @bmorris3 .

File: setup.cfg
Comment (pllim) [reply setup.cfg:64]: `romandeps` uses tox syntax, and tox is what is being used by the CI. What we have is copied over from how astropy do things.

If you want to run the test separately without tox, do you have to do this manually (no quotes):

`pip install -e .[roman]`

This syntax is standard in Python, not specific to Roman.

Also see https://github.com/spacetelescope/jdaviz/pull/1822#discussion_r1166870021

File: docs/imviz/import_data.rst
Comment (cshanahan1) [inline docs/imviz/import_data.rst:110]: you can also use the [romandeps] (or maybe just [roman]?) option you defined

File: jdaviz/configs/imviz/tests/utils.py
Comment (pllim) [reply jdaviz/configs/imviz/tests/utils.py:248]: Same comment as https://github.com/spacetelescope/jdaviz/pull/1822#pullrequestreview-1385545865 or is this something different?

Comment (bmorris3) [reply jdaviz/configs/imviz/tests/utils.py:248]: The function in `roman_datamodels` uses fixed pixel dimensions, which match the detector shape of Roman. For testing, it's convenient to make smaller arrays.

Comment (pllim) [reply jdaviz/configs/imviz/tests/utils.py:248]: p.s. LoL I was blind. I saw this URL but didn't cross my mind that we should just use the upstream version. Good catch!

File: tox.ini
Comment (pllim) [reply tox.ini:39]: Good question. Does it break a lot in dev? I know some pipeline merge their PRs precariously, resulting in broken dev version a lot. How does this package handle their merges?

Comment (pllim) [reply tox.ini:39]: I ask that because if it breaks a lot, it is pointless to test against their dev. But if not, then sure, we can.

File: docs/imviz/import_data.rst
Comment (pllim) [reply docs/imviz/import_data.rst:110]: Indeed! But the exact syntax changes depending on whether you want released version or dev version (editable or uneditable). How do we mention this without having to list out all the variants? Suggestion welcome.

File: jdaviz/configs/imviz/tests/utils.py
Comment (bmorris3) [reply jdaviz/configs/imviz/tests/utils.py:241]: It might be! `mk_level2_image` an `stnode.WfiImage` instance, right? If you run `rdd.open(wfi_image)`, would you get an `ImageModel`? (Testing this now).

File: jdaviz/configs/imviz/tests/utils.py
Comment (bmorris3) [reply jdaviz/configs/imviz/tests/utils.py:248]: To follow up: the link in the docstring corresponds to the version that I modified, and you can see that the detector shape can't be modified. On main, there's [a corresponding function that's more flexible](https://github.com/spacetelescope/roman_datamodels/blob/626e0da697614cefaef3b26d06fc3f528e2c7dab/src/roman_datamodels/maker_utils.py#L485)! I'll probably switch to using it instead.

File: docs/imviz/import_data.rst
Comment (bmorris3) [reply docs/imviz/import_data.rst:110]: This is true – but we might not want to suggest re-installing jdaviz just to install `roman-datamodels`.

**Review by kecnry — APPROVED** [review]
LGTM once comments from everyone else are addressed!  Squash & merge probably makes sense here.

File: jdaviz/configs/imviz/tests/utils.py
Comment (bmorris3) [reply jdaviz/configs/imviz/tests/utils.py]: Turns out this was an easy switch, see bd0ef9c.

File: jdaviz/configs/imviz/tests/utils.py
Comment (bmorris3) [reply jdaviz/configs/imviz/tests/utils.py:241]: Done! bd0ef9c

File: setup.cfg
Comment (bmorris3) [reply setup.cfg:64]: Right. With this PR, you can do
```
pip install jdaviz[roman]
pip install -e .[roman]
```
To get the roman deps. Before this PR, `[doc]` was already supported.

Specifying `-romandeps` in the tox args makes sure the correct packages are installed in the tox env for that job.

Comment (pllim) [reply setup.cfg:64]: Right, basically romandeps tells tox to go install the roman optional dependencies but using tox syntax. Or am I misunderstanding your comment?

File: CHANGES.rst
Comment (cshanahan1) [inline CHANGES.rst:15]: im assuming you just need to rebase which is why this extra stuff is in here

**Review by cshanahan1 — APPROVED** [review]

File: CHANGES.rst
Comment (bmorris3) [reply CHANGES.rst:15]: Yikes, I think this resulted from a bad rebase. I'll fix that now. Good catch!

**Comment by bmorris3** [comment]
Thanks all!

**Pull request merged**
